Adding FASTMath uber package and missing FASTMath packages

markcmiller86 opened the pull request:
Add FASTMath packages
- mesquite
- chombo (still problems with fortran name mangling)
- pumi
- phasta (still problems on osx)
- fastmath (uber package)

Modify packages
- boxlib to add variant for dims
- disable tools in moab

Modify setup-env.csh to only refer to $SYS_TYPE if its defined. Should probably be done for other shell types as well.

Chombo should probably be merged with...
https://github.com/drhansj/spack.git
branch develop
var/spack/repos/builtin/packages/chombo-xsdk/package.py

Finally, not sure I am doing this quite right. I am submitting this from same working copy I did an earlier pull request for ZFP from. Do I need to do something different to submit a pull request from that same working copy for entirely new work?

adamjstewart commented on share/spack/setup-env.csh:
```diff
@@ -41,7 +41,9 @@ if ($?SPACK_ROOT) then
+    if ($?SYS_TYPE) then
+        _spack_pathadd DK_NODE    "$_spack_share_dir/dotkit/$SYS_TYPE"
+        _spack_pathadd MODULEPATH "$_spack_share_dir/modules/$SYS_TYPE"
+    endif
```
I would put this in a separate PR with an explanation of why it's necessary. It isn't really related to FastMath.

markcmiller86 replied:
You're right it isn't related to fastmath. But, is the additional overhead of sep. PR for two lines of code really worth it? I didn't think so. That aside, as I am relatively new to git, how would one achieve this? I assume I need to somehow remove this file change from this PR and then cherry pick just this file change to create a new PR. Is that right?

adamjstewart replied:
Sometimes good old fashioned copy-n-paste is easier than messing with cherry picking. I would just undo these changes, push a commit, switch back to develop, create a new branch, and redo the change.

The problem with including unrelated commits in a PR is you never know when someone will disagree with the change. If they're in 2 PRs, neither change can hold up the other PR.

adamjstewart commented on var/spack/repos/builtin/packages/boxlib/package.py:
```diff
@@ -36,12 +36,14 @@ class Boxlib(CMakePackage):
 
     depends_on('mpi')
 
+    variant('dims', default=3, description='Number of spatial dimensions')
```
@alalazo Does this have to have a `values` argument?

markcmiller86 replied:
Based on mvapich example I saw, yes. At least once I switched to that, I got intended behavior. Is there a better approach?

adamjstewart replied:
I think that is the recommended approach. If there are a set number of dimensions allowed, e.g. 1-3, then you should use `values=(1, 2, 3)`. If it can be any integer, see `netcdf` for how that is handled.

adamjstewart commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,67 @@
+##############################################################################
+from spack import *
+import os
+from os.path import join as pjoin
```
`join_path`

markcmiller86 replied:
Not sure I follow. You mean I should use a function called join_path instead of pjoin?

adamjstewart replied:
Yeah. So I mentioned this in your other issue, but Spack has a `join_path` function that comes with `from spack import *`. It's just an alias to `os.path.join`.

markcmiller86 replied:
Ok, eliminated need for joining paths at all

adamjstewart commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,67 @@
+        dummy_tar_path = os.path.abspath(pjoin(os.path.split(__file__)[0]))
+        dummy_tar_path = pjoin(dummy_tar_path, "fastmath.tar.gz")
+        url      = "file://" + dummy_tar_path
+        return url
```
See how packages like `pgi` handle this. You might only need to define a `url` instead of `url_for_version`.

markcmiller86 replied:
Ok, the did use url_for_version but its simpler. So, now fastmath one is really simple

adamjstewart commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,67 @@
+
+    def install(self, spec, prefix):
+        make()
+        make('install')
```
Might want to extend `MakefilePackage` so you don't need an install method.

markcmiller86 replied:
Yeah, just doing that now. Didn't realize I hadn't started with that

markcmiller86 commented:
Ok, I backed out setup-env.csh changes from here. Will do another PR (at some point this week).
Updated as per other requests.

adamjstewart commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,58 @@
+##############################################################################
```
I didn't notice this the first time, but you probably shouldn't add the tarball to Spack. Users will have to download it manually and put it in the current directory or in a mirror.

markcmiller86 replied:
Thats not _really_ a tarball. Its basically empty, a dummy that is there to satisfy spack's need to have something to install. I don't like the idea of hosting  some dummy tarball in some mirror solely to affect that, especially since the tarball is a dummy thing. There is no FASTMath package per-se. Its just an uber install convenience promised as part of FASTMath project. Does that make sense?

adamjstewart replied:
Ah, I see. So `fastmath` is just a meta-package that installs other packages. There is a bit of precedence for this. I would take a look at the `xsdk` package that @BarrySmith wrote. It downloads a fake tarball (doesn't matter which) and writes a simple file to the installation directory.

Alternatively, you can ask users to run:
```
$ spack install --only=dependencies fastmath
```
instead so they don't have to worry about the `fastmath` installation.

markcmiller86 replied:
Well, I kinda figure its no worse than having a small patchfile there. I agree it _could_ be kinda confusing to other Spack develpers or anyone who goes wondering into fastmath package dir and sees that .tar.gz file.

FWIW, `spack install fastmath` is IMHO _much_ preferred to `spack install --only=dependencies fastmath`. So, I don't really like that solution. Any  way to have a package advertise --only=dependencies as its default behavior?

davydden replied:
I would do it the same way as xsdk

adamjstewart replied:
I agree that neither solution is particularly elegant. They're both hacks. The long term plan is to have some kind of a `MetaPackage` or `BundlePackage` base class that allows you to list dependencies without having to worry about fetching or installing. I'm not sure what the progress is on that. @citibeth might know more.

citibeth replied:
You should use #3133.  It needs to get merged, using it in this PR will be an incentive.

markcmiller86 commented:
Ok, so probably a silly question but from the git fork I had, where I had done work on develop with all these changes...I then did a git branch sys_type_fix, expecting I guess to create a clean copy of develop **before** my package changes. But, all I wound up doing was creating a branch that includes all my package changes for fastmath too. That isn't what we want. How do I cause git to create a branch from the main spack develop and NOT from my modified version of it? I am not even sure I am asking this question correctly. Sorry for naivety. If it was subversion...I would probably wind up creating a sep. working copy to do it.

tgamblin commented:
From where you're at now, I would just `git rebase -i develop` and remove the line for the commit you don't want.

markcmiller86 commented:
Well, I already removed the setup-env.csh fix @adamjstewart requested from this PR (at least I think I did).

What I wanna do now is create a new PR with just that change in it.

tgamblin commented:
Git checkout develop
Git checkout -b new-branch
Git cherry-pick <old commit with just that change>

Or make the change and commit again instead of that last step.

adamjstewart commented:
@tgamblin The problem is that @markcmiller86 committed these changes directly to his own develop branch. The way I usually do things is to keep my `develop` branch pure and only update with upstream LLNL develop. When I create a new PR, I create a new branch to commit changes to.

adamjstewart commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+    depends_on('hdf5', when='+hdf5')
+    depends_on('hdf5+mpi', when='+mpi+hdf5')
+
+    def edit(self, spec, prefix):
```
So this solution works fine, but a lot of the time, when you need to modify a `Make.defs` file, it's actually simpler to write a new one from scratch. See `elk`, `npb`, or `superlu` for examples. I know you have a deadline to get this in by the end of the week, so if this works as is and you don't want to change it I'm fine with it.

markcmiller86 replied:
I am not sure I agree its "simpler" to write one from scratch. Chombo has _a lot_ of variables defined (more than 60 I think). And, presumably, the list of variables varies with release. I think its more prudent for the build/install algorithm to take what is there and edit it than to try to create one from scratch.

adamjstewart replied:
I wasn't sure if you were overwriting every variable or just some. In that case, your solution works fine.

adamjstewart commented on var/spack/repos/builtin/packages/mesquite/package.py:
```diff
@@ -0,0 +1,55 @@
+        args += ['--with-mpi=%s' % self.spec['mpi'].prefix]
+        args += ['--enable-release']
+        args += ['--enable-shared']
+        return args
```
This is just personal preference, but I think this would look cleaner as:
```python
return [
    'CC=%s' % self.spec['mpi'].mpicc,
    'CXX=%s' % self.spec['mpi'].mpicxx,
    '--with-mpi=%s' % self.spec['mpi'].prefix,
    '--enable-release',
    '--enable-shared',
]
```

markcmiller86 replied:
Sure, I can change that. I think I cut-n-pasted from another package though. In most cases, I am looking at other code thats already been committed and simply cut-n-pasting that code and then adjusting it.

adamjstewart commented on var/spack/repos/builtin/packages/phasta/package.py:
```diff
@@ -0,0 +1,60 @@
+    def cmake_args(self):
+        spec = self.spec
+
+        cmake_args = std_cmake_args[:]
```
You don't need this line. `std_cmake_args` is already prepended to `cmake_args` in `cmake()` in `CMakePackage`.

markcmiller86 replied:
Again, thats from a cut-n-paste and I didn't bother to dig or think more deeply about it ;) I can fix this too.

BarrySmith commented:
@markcmiller86 Better IMHO to do a rebase -i to combine all the work you've done on building a pull request into a single commit before issuing a pull request. This makes the pull request more coherent; if a branch with a bunch of commits is in the pull request the resulting blob introduced into the develop branch doesn't have the connections between the commits in the branch and hence is difficult to understand later. Similarly you can do this again after you've resolved any requests for changes.

markcmiller86 commented:
@BarrySmith ...I obviously still have much to learn about using git (and unlearning overloaded subversion verbiage that often prevents my understanding of git).

BarrySmith commented:
Learning basic git is easy, unlearning subversion is hard :-)

adamjstewart commented:
Don't worry too much about how many commits are in a PR. We squash all PRs into a single commit before merging anyway.

davydden commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+    patch('hdf5-16api.patch', when='@3.2', level=0)
+    patch('Make.defs.local.template.patch', when='@3.2', level=0)
+
+    depends_on('lapack')
```
depends_on('blas`) is missing

adamjstewart replied:
This package might not depend on BLAS.

adamjstewart replied:
Oh, nvm, I see it's used below.

davydden replied:
> This package might not depend on BLAS.

every package which uses `lapack` needs `blas` by _construciton_ :smile:

markcmiller86 replied:
Hmm. Isn't that taken care of because lapack itself depends on blas?

davydden replied:
> Isn't that taken care of because lapack itself depends on blas?

I am not sure, it does depend indeed, but theoretically you may have a package which provided both lapack and blas (e.g. `atlas`), but when you write `depends_on('lapack')`, i think `blas` will not be a part of the dependency as `lapack` virtual requirement will be fulfilled by a single package, i.e. `atlas`. Long story short -- i think it's safer to write both.

davydden commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+            'lapackincflags = -I%s' % spec['lapack'].prefix.include)
+        defs_file.filter(
+            '^\s*#\s*syslibflags\s*=\s*',
+            'syslibflags= -L%s -llapack -lblas' % spec['lapack'].prefix.lib)
```
please, don't hardcode lib names. see how it is used in any other package which needs blas/lapack. 
You need to do
```
lapack_blas = spec['lapack'].libs + spec['blas'].libs
```
and then use it as
```
lapack_blas.ld_flags
```

markcmiller86 replied:
Oh, sheesh. I had forgotten I did that. Fixing now.

markcmiller86 replied:
So, for better or worse, when I look at thinks like superlu-dist for example, it really isn't all that obvious from the package.py file what those spack variables (or objects) are doing...

`
        lapack_blas = spec['lapack'].libs + spec['blas'].libs
.
.
.
            'BLASDEF      = -DUSE_VENDOR_BLAS',
            'BLASLIB      = %s' % lapack_blas.ld_flags,
`

Is that going to wind up setting BLASLIB to `-L/some/lapack/path -L/some/blas/path -llapack -lblas` or is it going to produce just the lower-case l options (which is ordinarily what I think the LDFLAGS is _only_ used for uppercase -L options and LDLIBS is used for lower case.

davydden replied:
it's quiet well documented, look at https://github.com/LLNL/spack/blob/develop/lib/spack/llnl/util/filesystem.py#L744  doc page does not exist, for some reason http://spack.readthedocs.io/en/latest/llnl.util.rst.html?highlight=librarylist#llnl.util.filesystem.LibraryList

davydden replied:
@markcmiller86 what i wrote above will do what you want. Just try and see.

adamjstewart replied:
http://spack.readthedocs.io/en/latest/llnl.util.html#module-llnl.util.filesystem

davydden commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+            'syslibflags= -L%s -llapack -lblas' % spec['lapack'].prefix.lib)
+
+        # Compilers and Compiler flags
+        defs_file.filter('^\s*#\s*CXX\s*=\s*', 'CXX = c++')
```
`CXX=%s` % spack_cxx

markcmiller86 replied:
fixing now

davydden commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+
+        # Compilers and Compiler flags
+        defs_file.filter('^\s*#\s*CXX\s*=\s*', 'CXX = c++')
+        defs_file.filter('^\s*#\s*FC\s*=\s*', 'FC = f90')
```
`FC=%s` % spack_fc

markcmiller86 replied:
fixing now

davydden commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+            'MPI = %s' % ('TRUE' if '+mpi' in spec else 'FALSE'))
+        defs_file.filter(
+            '^\s*#\s*DIM\s*=\s*',
+            'DIM = %d' % int(spec.variants['dims'].value))
```
please add an assert at the beginning w.r.t. the allowed values of `dims`, probably 1-3 only.

adamjstewart replied:
There is finally support for that in multi-value variants. This would look like `values=(1, 2, 3)`.

davydden replied:
yes, that's better way of doing it.

markcmiller86 replied:
Ok, will do

markcmiller86 replied:
Hmm. Just asking now...mvapich2 has ch3_rank_bits that can have two values; 32 or 64.

However, there is nothing either documents that set of acceptable values or asserts them either.

My practice is generally to go find how another package did it and copy that pattern. So, what you see in my PR here is (mostly) a reflection of things like that. 

My question is, why is this PR then getting held up regarding these minute details when there is already precedent in the existing set of packages? Is there some style guide I should have read to have known better ahead of time. Is there any effort to go back and re-code existing packages to reflect new style/policies in use of Spack? Just curious

markcmiller86 replied:
@adamjstewart ...I am not finding any instance/example of a 'values=' statement anywhere in the packages.

`/Users/miller86/spack/spack/var/spack/repos/builtin/packages
[scratlantis:repos/builtin/packages] miller86% grep 'values=' */*.py
[scratlantis:repos/builtin/packages] miller86% 
`

citibeth replied:
> My question is, why is this PR then getting held up

This PR is only 20 hours old.  It is getting expediated review.  The kinds of changes being asked for here are typical of most PRs containing new packages with strange/complex builds.  There are PRs that have been held up many months.

> regarding these minute details when there is already precedent in the existing set of packages? 

Because nobody fixed the previous precedent.  Which is not a good idea to let more stuff we don't want get through.

> Is there some style guide I should have read to have known better ahead of time.

We've talked about making one, but so far no. The fear is it would be so detailed it chases off the average first-committer.  So far, we just have humans review them and bring up the issues as they arise.

> Is there any effort to go back and re-code existing packages to reflect new style/policies in use of Spack?

Yes.

adamjstewart replied:
@markcmiller86 The reason is that multi-valued variant support was just drastically changed in #2386. For better or for worse, if you don't specify what values are allowed, things crash. @alalazo is fixing this in #4105. If you look at `mvapich2` from the latest develop branch, you'll see that it now sets `values` appropriately.

markcmiller86 replied:
Ok, sounds like I need to do a git pull upstream to see that?

adamjstewart replied:
Yep! You might run into problems because you are working on your `develop` branch. Let me know if you do. For future reference, it's probably better to create a separate branch for any changes.

markcmiller86 replied:
@adamjstewart ...so, curious...why are the _values_ in mvapich2 string valued for ch3_rank_bits? Thats an integer valued option.

davydden commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+                'HDFINCFLAGS = -I%s' % spec['hdf5'].prefix.include)
+            defs_file.filter(
+                '^\s*#\s*HDFLIBFLAGS\s*=.*',
+                'HDFLIBFLAGS = -L%s -lhdf5' % spec['hdf5'].prefix.lib)
```
```
'HDFLIBFLAGS = %s' % spec['hdf5'].libs.ld_flags
```
should work just fine.

markcmiller86 replied:
fixing now

davydden commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+                    'HDFMPIINCFLAGS = -I%s' % spec['hdf5'].prefix.include)
+                defs_file.filter(
+                    '^\s*#\s*HDFMPILIBFLAGS\s*=.*',
+                    'HDFMPILIBFLAGS = -L%s -lhdf5' % spec['hdf5'].prefix.lib)
```
same here

markcmiller86 replied:
fixing now.

davydden commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,58 @@
+    depends_on('zoltan')
+
+    def url_for_version(self, version):
+        return "file://%s/fastmath.tar.gz" % os.getcwd()
```
do you not need to have `install` for this meta-package, i.e. 
```
def install(self, spec, prefix):
        # Prevent the error message
        #      ==> Error: Install failed for <package>.  Nothing was installed!
        #      ==> Error: Installation process had nonzero exit code : 256
        with open(os.path.join(spec.prefix, 'bundle-package.txt'), 'w') as out:
            out.write('This is a bundle\n')
            out.close()
```

markcmiller86 replied:
Well, I guess I choose a different route than @BarrySmith on xsdk. So, yes, I need the _install_. If there is _supposed_ to be an adopted convention for this, can we do something to make it part of Spack instead of expecting package writers to know? Also, I think I've been told to use join_path and this code you reference here doesn't use it.

davydden replied:
i think there is no convention, i just though that you **need** `install()`, otherwise the bundled package will fail.

adamjstewart replied:
`join_path == os.path.join`, so it doesn't really matter.

There isn't really an established convention because #3133 will make previous conventions meaningless. In the meantime, you can use whatever you want, although @BarrySmith's convention doesn't rely on any additional files or Makefiles.

markcmiller86 replied:
@davydden ...sorry I misunderstood your inquiry regarding need for an install method. I don't think MakefilePackage **requires** an install method. By setting to MakefilePackage, I believe default install method results in 'make install' being invoked. I constructed the dummy fastmath.tar.gz file to simply do that. If I am wrong about that, I wouldn't have expected my install to work and result in the README file getting installed to the fastmath package install point.

davydden replied:
> I constructed the dummy fastmath.tar.gz file to simply do that.

i understand, but i still don't like adding a binary file to Git VCS only for the sake of dummy installation. Especially give the fact that it's going to be removed in a week or so.

davydden replied:
p.s. originally i did not notice that you are adding this binary file, thus i was surprised that things are installing at all.

davydden commented on var/spack/repos/builtin/packages/pumi/package.py:
```diff
@@ -0,0 +1,64 @@
+    url      = "https://github.com/SCOREC/core.git"
+
+    version('master', git='https://github.com/SCOREC/core.git',
+        branch='master')
```
keep on one line, please

markcmiller86 replied:
will fix

davydden replied:
> will fix

thanks.

p.s. IMO you do NOT really have to reply to each comment that you are fixing it. When you push changes where this line is modified, then the whole comment will be collapsed/closed. It is then understood that you changed this line according to the comment.

davydden commented on var/spack/repos/builtin/packages/pumi/package.py:
```diff
@@ -0,0 +1,64 @@
+    homepage = "https://www.scorec.rpi.edu/pumi"
+    url      = "https://github.com/SCOREC/core.git"
+
+    version('master', git='https://github.com/SCOREC/core.git',
```
`master` -> `develop`

citibeth replied:
This is not checksummable, needs to be fixed.  You need at least one checksummable version in every package.  If there is no tarball release, just add a version '0.0.1' based on the latest git hash in master.

markcmiller86 replied:
I've been told by Pumi developers to use master. Have you heard otherwise?

davydden replied:
> I've been told by Pumi developers to use master. Have you heard otherwise?

i meant how you call the version, i.e. first `master`, not the branch name!

citibeth replied:
@davydden  means you want to change to:
```
+    version('develop', git='https://github.com/SCOREC/core.git',
+        branch='master')
```

davydden replied:
you need to call it `develop`, otherwise it will NOT be compared correctly with other releases, such as `x.y`.

adamjstewart replied:
What @davydden means is that Spack uses version numbers for things like sorting. If I ask for `pumi@2.5:`, Spack needs to know whether the version is greater than or less than 2.5. Right now, `develop` is considered to be _higher_ than any other numbered version, while all other non-numeric strings are considered to be _less_ than any numbered versions. I personally view this as a bug and think `master ` and `develop` should both be treated as greater than all numeric versions, but that started a long argument that spans multiple issues and fell to the wayside as there were bigger fish to fry. For what it's worth, I personally add `master` versions to Spack and don't care much for version sorting if there's only one version.

markcmiller86 replied:
ah, ok...got it. wil fix

davydden commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,58 @@
+import os
+
+
+class Fastmath(MakefilePackage):
```
if it's a bundle, no need to make it a `MakefilePackage`, just derive from `Package` class.

citibeth replied:
Use #3133 and derive from `BundlePackage`.

tgamblin replied:
#3133 is not merged.  Don't use it.

citibeth replied:
@tgamblin Can you give a concrete suggestion on what to do without #3133 available?  And how much work would that be compared to finishing up #3133 and getting it merged (which should not be too hard).

adamjstewart replied:
I think the way this package does it or the way `xsdk` does it would be fine for now. We can update both once #3133 is merged.

markcmiller86 replied:
That is part of my choice in having a dummy tarball that supports a 'make install' target.

citibeth replied:
Can you please put the string "BundlePackage" somewhere in a comment in the file?  We will grep for it after #3133 is merge.

davydden replied:
> That is part of my choice in having a dummy tarball that supports a 'make install' target.

oh, i see. Let's NOT commit a dummy binary file to Github only for the sake of making something install. You can achieve the same via very few lines of code like in `xsdk` package.

adamjstewart replied:
To be fair, this solution doesn't rely on being able to download some random tarball from online and ignore it completely, so I actually find it a bit more elegant than @BarrySmith's solution. Anyway, it doesn't matter in the long run. #3133 looks pretty solid to me. I'm sure we can get it merged in the next couple weeks.

citibeth replied:
Instead of committing a dummy tarball to Spack, download it from this repo: 
 https://github.com/citibeth/dummy

And subclass `Package`.  That is what I did before #3133.

davydden replied:
> To be fair, this solution doesn't rely on being able to download some random tarball from online and ignore it completely

that's true, but adding a binary to git VCS is not elegant either. Since  #3133 will be merged soon, let's not spoil the git history of Spack with temporary binaries.

markcmiller86 replied:
Will consider better options. I don't need to gzip it though. Would that remove its _binary_ flavor?

adamjstewart replied:
Spack unzips it after downloading automatically. Although you can add `expand=False` to the `version()` directive to prevent this.

tgamblin replied:
I'll work on getting #3133 merged and supporting this better.  I agree it's a pain.

citibeth commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+       requires human registration process."""
+
+    homepage = "https://anag-repo.lbl.gov/svn/Chombo/release"
+    url      = "https://anag-repo.lbl.gov/svn/Chombo/release/3.2"
```
We need to contact the original Combo authors before doing this.

markcmiller86 replied:
I've been in contact with them. I work with them on FASTMath. What is your concern?

citibeth commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+    homepage = "https://anag-repo.lbl.gov/svn/Chombo/release"
+    url      = "https://anag-repo.lbl.gov/svn/Chombo/release/3.2"
+
+    version('3.2', svn='https://anag-repo.lbl.gov/svn/Chombo/release/3.2')
```
This version is not checksummable.  It will therefore not work as the default Spack version.

markcmiller86 replied:
Ok, I've had a telephone call with Brain Van Straalen regarding Chomb in Spack. He will be sending me a github path, the same one I think 3133 is currently using and it'll become a github download rather than a less save, unchecksumed, svn download. Satisfactory?

citibeth commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+        # Set fortran name mangling in Make.defs
+        defs_file = FileFilter('./lib/mk/Make.defs')
+        defs_file.filter('^\s*#\s*cppcallsfort\s*=\s*',
+                         'cppcallsfort = -DCH_FORT_UNDERSCORE')
```
Which Fortran+C compiler combos will this work or not work on?  Why don't we use existing name mangling detection provided by CMake or Autotools?  We don't have to move the whole build to CMake, just ask it about name mangling.

markcmiller86 replied:
So, far, just OSX and one LLNL system I've tested on. Still working on fortran name mangling for this

citibeth commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,130 @@
+        # Unconditional settings
+        defs_file.filter('^\s*#\s*DEBUG\s*=\s*', 'DEBUG = FALSE')
+        defs_file.filter('^\s*#\s*OPT\s*=\s*', 'OPT = TRUE')
+        defs_file.filter('^\s*#\s*PIC\s*=\s*', 'PIC = TRUE')
```
Shouldn't DEBUG, OPT and PIC be linked up to Spack variants?

citibeth commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,58 @@
+
+    version('1.0', 'cd5006bbe9a39c1bf252b9b8b5e5bdcd')
+
+    depends_on('boxlib dims=3')  # how do we say 3D boxlib?
```
@alalzo Is this the right syntax for setting multi-valued variants?

adamjstewart replied:
Yes.

tgamblin replied:
Yes

citibeth commented on var/spack/repos/builtin/packages/mesquite/package.py:
```diff
@@ -0,0 +1,56 @@
+            'CC=%s' % self.spec['mpi'].mpicc,
+            'CXX=%s' % self.spec['mpi'].mpicxx,
+            '--with-mpi=%s' % self.spec['mpi'].prefix,
+            '--enable-release',
```
Release and shared should be hooked up to Spack variants.

markcmiller86 replied:
What? Is this true across all 1400 packages? I choose to minimize variant proliferation at this point. It doesn't mean they can't be added later. I just don't see the need to complicate things until I/we know we have to.

davydden replied:
> It doesn't mean they can't be added later. 

I agree with @markcmiller86 , there is no pressing need to do it **now**, in the first version of the package.

citibeth commented on var/spack/repos/builtin/packages/moab/package.py:
```diff
@@ -67,13 +68,35 @@ def install(self, spec, prefix):
+            '--disable-mbquality',
+            '--disable-ahf',
+            '--disable-mbumr',
+            '--disable-imesh',
```
Why all these disables?  How will it affect other Moab users?

markcmiller86 replied:
Because --disable-tools (which was _supposed_ to have that effect) was apparently missing some...because the tools are needed to install the library and because they create additional dependencies I don't want to add at this time. I work with MOAB developers too. I have adopted a strategy of minimal installs for which we can slowly add functionality.

adamjstewart replied:
I agree with disabling things that we don't have Spack packages for. With most build systems, if you don't specify, it will install things if the dependencies are present and won't if they aren't. This results in unreproducible builds: something works fine for one user but not for another. It's better to explicitly enable/disable every variant that behaves this way.

markcmiller86 replied:
Well, to be fair, there probably _are_ spack packages for any of the dependencies the tools introduce. But, I just don't think users of MOAB need those tools. Developers of MOAB would though.

citibeth replied:
> . With most build systems

You mean, with Autotools.  CMake has tried to educate the world about resisting the temptation to auto-configure based on what's available.

citibeth commented on var/spack/repos/builtin/packages/moab/package.py:
```diff
@@ -58,6 +58,7 @@ class Moab(Package):
     depends_on('mpi')
     depends_on('hdf5+mpi')
     depends_on('netcdf+mpi', when='+netcdf')
+    depends_on('metis')
```
Why the added dependency?  Does it change the way Moab builds?

markcmiller86 replied:
My conclusion was that it needs parmetis in parallel. But, when it needs parmetis, it also needs metis. I don't think there is a serial configuration that needs metis without parmetis though. Does that make sense?

davydden replied:
> But, when it needs parmetis, it also needs metis.

different from `blas`/`lapack` discussion, here you do not need to add `metis`, It will be brought through `parmetis` dependency. During configure all you do is `--with-parmetis=%s' % spec['parmetis'].prefix`, and `metis` is not used anyhow. So i think this line can go.

citibeth commented on var/spack/repos/builtin/packages/moab/package.py:
```diff
@@ -58,6 +58,7 @@ class Moab(Package):
     depends_on('mpi')
     depends_on('hdf5+mpi')
     depends_on('netcdf+mpi', when='+netcdf')
```
This also probably needs a plain `depends_on('netcdf')`

markcmiller86 replied:
ok. will fix

markcmiller86 replied:
Hmm. More I think about this...I conclude it would need a plain `depends_on(netcdf)` _only_ if building with and without MPI is supported. MPI dependence is UNconditiional here. Maybe I should make at least that variant here?

davydden replied:
since `mpi` is unconditional, i think this is a perfectly fine line. At this point, IMO there is no need making `mpi` a variant and then you may or may not need `depends_on('netcdf')`.  I would leave it as-is.

citibeth commented on var/spack/repos/builtin/packages/phasta/package.py:
```diff
@@ -0,0 +1,59 @@
+
+    depends_on('mpi')
+
+    def cmake_args(self):
```
Has this code been tested?  It looks like `cmake_args` is a method, not a list.

markcmiller86 replied:
Copieed it from another package. I have tested it on OSX.

adamjstewart replied:
Which package? I think you need to add `cmake_args = []` to the beginning.

citibeth commented on var/spack/repos/builtin/packages/phasta/package.py:
```diff
@@ -0,0 +1,59 @@
+            '-DCMAKE_Fortran_COMPILER=%s' % spec['mpi'].mpifc,
+        ])
+
+        return cmake_args
```
Try just:
``` 
return [
+            '-DPHASTA_USE_MPI=ON',
 +            '-DPHASTA_BUILD_CONVERTERIO=OFF',
 +            '-DPHASTA_BUILD_ACUSTAT=OFF',
 +            '-DPHASTA_BUILD_M2N=OFF',
 +            '-DPHASTA_BUILD_M2NFixBnd=OFF',
 +            '-DPHASTA_USE_LESLIB=OFF',
 +            '-DPHASTA_USE_PETSC=OFF',
 +            '-DPHASTA_USE_SVLS=ON',
 +            '-DPHASTA_INCOMPRESSIBLE=ON',
 +            '-DPHASTA_COMPRESSIBLE=ON',
 +            '-DCMAKE_C_COMPILER=%s' % spec['mpi'].mpicc,
 +            '-DCMAKE_CXX_COMPILER=%s' % spec['mpi'].mpicxx,
 +            '-DCMAKE_Fortran_COMPILER=%s' % spec['mpi'].mpifc]
```

markcmiller86 replied:
Umm. No. I like it the way it is, especially if may be revised with more complex conditional logic later.

tgamblin commented:
@citibeth: while I would like to merge #3133, it's just going to complicate merging this one to have @markcmiller86 integrate it.  we can integrate it after this one goes in.

markcmiller86 commented:
@citibeth ...chombo not beeing checksumable...GitHub didn't provide place for me to comment directly so am commnting here. its a git download. Everytime I've run spack it tells me checksums are not needed for git downloads. So, I am confused by your statement.

davydden commented on var/spack/repos/builtin/packages/phasta/package.py:
```diff
@@ -0,0 +1,59 @@
+    def cmake_args(self):
+        spec = self.spec
+
+        cmake_args.extend([
```
Probably you should do the following:
before this line add
```
args = []
```
and then do
```
args.extend([
  blabla
])
return args
```

citibeth replied:
Or just return the value without pointlessly putting it in a variable first.

citibeth commented:
@markcmiller86 See #2761 and search for "checksummable."  To solve your problem, you need to use something like:
```
version('0.0.1', git='https://github.com/example-project/example.git',
        commit='9d38cd4e2c94c3cea97d0e2924814acc')
```
Choose the latest commit hash on the `master` branch.  By using a commit hash, Spack is guaranteed that the version will never change (and Git checkout by commit hash IS checksummable).  You can also keep what you have here, something like:
```
    version('develop', git='https://github.com/PHASTA/phasta.git',
        branch='master')
```
When users ask for your package without anything special, they will get the `0.0.1` version, which corresponds to a specific (checksummable) hash.  When they want the latest-greatest (and possibly compromised), they can ask for `@develop`.

citibeth commented:
This stuff is important to prevent users from unwittingly installing code that Spack has no way to verify the safety of.

tgamblin commented:
@markcmiller86: @citibeth just wants a commit id associated with the git pull.  It looks like you could do that for SCOREC.

I am not sure what to do about Chombo, as it looks like the only way it is distributed is over `svn`.  Requiring the svn URL to be `https` probably as good as we can do there.

tgamblin commented:
> This stuff is important to prevent users from unwittingly installing code that Spack has no way to verify the safety of.

@citibeth: note that we actually don't verify the safety of anything.  We try to ensure that contributors are fetching from reasonable sites when we accept pull requests, and we have discussed automatically checking the validity of checksums with Travis.  The checksum ensures that what users get is what was we checked in the PR, but that's it.  I'm hesitant to call that "safety".

citibeth commented:
If we can't verify for chombo I say we punt.  At least we tried...
On May 4, 2017 2:11 PM, "Todd Gamblin" <notifications@github.com> wrote:

> @markcmiller86 <https://github.com/markcmiller86>: @citibeth
> <https://github.com/citibeth> just wants a commit id associated with the
> git pull. It looks like you could do that for SCOREC.
>
> I am not sure what to do about Chombo, as it looks like the only way it is
> distributed is over svn. Requiring the svn URL to be https probably as
> good as we can do there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/4120#issuecomment-299265662>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd2K_MiKdSyJpi0VwcqcY84wbWpUmks5r2hTUgaJpZM4NQC1L>
> .
>

tgamblin commented:
@citibeth @markcmiller86: It would be nice if the chombo install pointed to a specific revision, though.  I am wary of having moving targets as the default fetched version.

tgamblin commented:
@citibeth @markcmiller86: Ignore my prior comment -- it does!

citibeth commented:
Super!
On May 4, 2017 5:15 PM, "Mark C. Miller" <notifications@github.com> wrote:

> *@markcmiller86* commented on this pull request.
> ------------------------------
>
> In var/spack/repos/builtin/packages/chombo/package.py:
>
> > +from spack import *
> +from shutil import copyfile
> +import glob
> +
> +
> +class Chombo(MakefilePackage):
> +    """The Chombo package provides a set of tools for implementing finite
> +       difference and finite-volume methods for the solution of partial
> +       differential equations on block-structured adaptively refined
> +       logically rectangular (i.e. Cartesian) grids. Note: Chombo download
> +       requires human registration process."""
> +
> +    homepage = "https://anag-repo.lbl.gov/svn/Chombo/release"
> +    url      = "https://anag-repo.lbl.gov/svn/Chombo/release/3.2"
> +
> +    version('3.2', svn='https://anag-repo.lbl.gov/svn/Chombo/release/3.2')
>
> Ok, I've had a telephone call with Brain Van Straalen regarding Chomb in
> Spack. He will be sending me a github path, the same one I think 3133 is
> currently using and it'll become a github download rather than a less save,
> unchecksumed, svn download. Satisfactory?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/4120#discussion_r114889643>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd3AI9NGjpgDiP9pznm-nF6AKYxBHks5r2j_-gaJpZM4NQC1L>
> .
>

markcmiller86 commented:
That's not working for chombo now. Its not installing any libs.

davydden commented on var/spack/repos/builtin/packages/phasta/package.py:
```diff
@@ -0,0 +1,61 @@
+    def cmake_args(self):
+        spec = self.spec
+
+        cmake_args = [
```
rename to `args` or `options`.

davydden commented on var/spack/repos/builtin/packages/pumi/package.py:
```diff
@@ -0,0 +1,66 @@
+    def cmake_args(self):
+        spec = self.spec
+
+        cmake_args = std_cmake_args[:]
```
same here, (i) use `args`/`options` (ii) you do NOT need to add `std_cmake_args`, this is done by CMake class automatically.

markcmiller86 commented:
I know what happened with Chombo build...I changed version number to reflect github version and didn't update the patch statement's when directives.

markcmiller86 commented:
I think I've addressed everything except the numerical variant settings. Still working on git commands to pull most recent stuff into my local git repo.

markcmiller86 commented:
Ok, I think I've taken care of all requests now. Thanks for hand holding through git issues

citibeth commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,142 @@
+                    '^\s*#\s*HDFMPILIBFLAGS\s*=.*',
+                    'HDFMPILIBFLAGS = %s' % spec['hdf5'].libs.ld_flags)
+
+        return
```
No need for this line.

citibeth commented on var/spack/repos/builtin/packages/mesquite/package.py:
```diff
@@ -0,0 +1,56 @@
+            '--enable-release',
+            '--enable-shared',
+        ]
+        return args
```
Can you eliminate the variable `args` altogether?  Just do this:
```
def configure_args(self):
+        return [
+            'CC=%s' % self.spec['mpi'].mpicc,
+            'CXX=%s' % self.spec['mpi'].mpicxx,
+            '--with-mpi=%s' % self.spec['mpi'].prefix,
+            '--enable-release',
+            '--enable-shared',
+        ]
```

citibeth commented on var/spack/repos/builtin/packages/phasta/package.py:
```diff
@@ -0,0 +1,61 @@
+            '-DCMAKE_Fortran_COMPILER=%s' % spec['mpi'].mpifc,
+        ]
+
+        return args
```
Can you get rid of the pointless `args` variable?

davydden replied:
same as below.

citibeth commented on var/spack/repos/builtin/packages/pumi/package.py:
```diff
@@ -0,0 +1,65 @@
+            '-DCMAKE_CXX_COMPILER=%s' % spec['mpi'].mpicxx,
+        ]
+
+        return args
```
Get rid of pointless `args` variable?

davydden replied:
i think @markcmiller86 prefers to keep it this way as it is easier to introduce variants later.

citibeth commented:
Question: How did you resolve the Fortran name mangling issue?  Will it work for all Spack-supported compilers?

markcmiller86 commented:
@citibeth...

> Question: How did you resolve the Fortran name mangling issue? Will it work for all Spack-supported compilers?

Not yet. To truly resolve this _automatically_, I think the package would need to detect it. I think if such detection were generally available now, it should be part of spack internals and not confined to one of its packages (many packages could probably benefit). I think the CMake logic in TRIBITS (which is what Trilino's CMake world uses), has the essential stuff to do this. But, in TRIBITS, I think that functionality is spread across a number of cmake files and not easily extricated as a standalone capability -- though I think some other teams have taken to attempting that.

So, at present I am leaning to punting and making it into a variant that a user _must_ choose a value for. Thats not too helpful but would resolve it in the absence of an automated solution.

One other thing I've considered is a simple hello-world Fortran shared library that gets compiled and then called from a C caller and C++ caller via dl. dlopen and introspection of the hello-world dynamic library would reveal how the Fortran symbol appears to C/C++ callers. That information could then be used within package.py file to prime name mangling setting.

adamjstewart commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,142 @@
+    patch('hdf5-16api.patch', when='@3.2', level=0)
+    patch('Make.defs.local.template.patch', when='@3.2', level=0)
+
+    depends_on('blas')  # if req'd whenever lapack is, y !Spack enforce it
```
BLAS isn't really required whenever LAPACK is. They are usually both provided by the same package. But in this case, you are using both BLAS and LAPACK, so it's best to explicitly declare this dependency. Looks good now.

markcmiller86 replied:
Hmm. Well, @davydden requested I add depends_on('blas'). It was not there originally.

adamjstewart replied:
Yes, it looks good now. You can probably remove this comment.

adamjstewart commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,59 @@
+    version('1.0', 'cd5006bbe9a39c1bf252b9b8b5e5bdcd')
+
+    # BundlePackage
+    depends_on('boxlib dims=3')  # how do we say 3D boxlib?
```
This is the correct way to say it. You can remove this comment.

adamjstewart commented on var/spack/repos/builtin/packages/mesquite/package.py:
```diff
@@ -0,0 +1,56 @@
+    homepage = "https://software.sandia.gov/mesquite"
+    url      = "https://software.sandia.gov/mesquite/mesquite-2.3.0.tar.gz"
+
+    version('2.99',  '92b94167981bb8fcd59b0f0f18fbab64')
```
It looks like `2.99` is the current development version and `2.3.0` is the latest release. Right now, Spack will choose `2.99` as the default version to use. If you want it to use `2.3.0` by default instead, you would add `preferred=True` to the `version()` directive. I don't know enough about Mesquite to know which version should be preferred though.

markcmiller86 replied:
2.99 has been the version I've used for the last 4-5 years. I don't think any development has been going on on mesquite for quite a while. But, I wonder if 2.99 is at risk of becoming a moving target in the future? Would it be better to just re-label it as 'develop' instead of 2.99?

adamjstewart replied:
That depends. I think it's fine to keep it as 2.99 for now. If we notice that they frequently update the tarball without changing the version number, we could consider calling it `develop`.

citibeth commented:
> > Question: How did you resolve the Fortran name mangling issue? Will it work for all Spack supported compilers?
> Not yet. To truly resolve this automatically, I think the package would need to detect it.

See #4136

That's the "easy" part.  The "hard" part is using that information effectively in Chombo.  I looked at Chombo's `FortranNameMacro.H` file, which is would would need to be patched/modified/updated.  Here's the beginning...

```
#ifdef CH_FORT_UPPERCASE
   #ifdef CH_FORT_UNDERSCORE
      #define FORTRAN_BASENAME( NAME ,name ) NAME ## _
      #ifdef CH_USE_DOUBLE
         #define LAPACK(NAME, name) d ## NAME ## _
      #else
         #define LAPACK(NAME, name) s ## NAME ## _
      #endif
   #else
      #define FORTRAN_BASENAME( NAME ,name ) NAME
      #ifdef CH_USE_DOUBLE
         #define LAPACK(NAME, name) d ## NAME
      #else
         #define LAPACK(NAME, name) s ## NAME
      #endif
   #endif
#else
   #ifdef CH_FORT_UNDERSCORE
      #define FORTRAN_BASENAME( NAME ,name ) name ## _
      #ifdef CH_USE_DOUBLE
         #define LAPACK(NAME, name) d ## name ## _
      #else
         #define LAPACK(NAME, name) s ## name ## _
      #endif
   #else
      #define FORTRAN_BASENAME( NAME ,name ) name
      #ifdef CH_USE_DOUBLE
         #define LAPACK(NAME, name) d ## name
      #else
         #define LAPACK(NAME, name) s ## name
      #endif
   #endif
#endif
```

Assuming we `#include FCMangle.h`, that That segment would have to change to:

```
      #define FORTRAN_BASENAME( NAME ,name ) FC_GLOBAL(name,NAME)
      #ifdef CH_USE_DOUBLE
         #define LAPACK(NAME, name) d ## FC_GLOBAL(name,NAME)
      #else
         #define LAPACK(NAME, name) s ## FC_GLOBAL(name,NAME)
      #endif
```

The overall result would be that `FrotranNameMacro.H` would be 1/4 the size it is now.

> I think if such detection were generally available now, it should be part of spack internals and not confined to one of its packages (many packages could probably benefit).

Well... maybe.  In general, Spack's job is to find and arrange dependencies, and provide a stage for packages to build.  Well-behaved packages should configure and build themselves.  Spack is far from the best tool available for doing the actual build or install.  That's what CMake is for.

It is for this reason that I believe we should encourage upstream authors to move away from ad-hoc manually edited Makefiles.  They are a *lot* more work to use in Spack (as you've seen in this PR), and the Spack build often ends up doing things that really should be done by CMake.

At the end of the day, if we go to a lot of Spack-specific work "improving" a build by adding additional auto-detection, or actually installing the output (see `googletest/package.py`), this effort does nothing to improve non-Spack builds of the upstream package.  But sometimes we have to go down that route as well.  I just believe it is our Plan B.  Plan A should always be to improve the upstream pacakge's build, so it benefits everybody using the package.

> So, at present I am leaning to punting and making it into a variant that a user must choose a value for. Thats not too helpful but would resolve it in the absence of an automated solution.

No.  Leaving the problem unsolved is a better approach, IMHO.  Eventually someone who wants to build on a compiler that name-mangles differently from your chosen compiler will maybe have more motivation to finish the task.  Just leave the URLs for #4120 and #4136 in `chombo/package.py` comments, so people can pick up where we've left off.

> One other thing I've considered is a simple hello-world Fortran shared library that gets compiled and then called from a C caller and C++ caller via dl. dlopen and introspection of the hello-world dynamic library would reveal how the Fortran symbol appears to C/C++ callers. That information could then be used within package.py file to prime name mangling setting.

This is exactly the kind of stuff that CMake and Autotools already do.  So we don't want to reinvent that wheel.

markcmiller86 commented:
> @citibeth At the end of the day, if we go to a lot of Spack-specific work "improving" a build by adding additional auto-detection, or actually installing the output (see googletest/package.py), this effort does nothing to improve non-Spack builds of the upstream package. But sometimes we have to go down that route as well. I just believe it is our Plan B. Plan A should always be to improve the upstream pacakge's build, so it benefits everybody using the package.

I'm impressed you dug into the bowels of chombo to find its name mangling logic. It actually does a ton of code generation during its build phase too. I agree here that its best to have package developers more informed about how to handle various issues. But, they've got something thats been working for them that I am sure they would prefer not to have to go back and change now. So, I kinda think we are stuck with what such packages do. I do work with developers of Chombo infrequently throughout the year and can feedback to them information.

That said, ineffective methods for fortran name mangling proliferate throughout the HPC world (as do shared/static debug/release, 32/64 bit indexing, LDFLAGS/LIBS (maybe a few other tall poles I can't recall right now). I don't think anyone has written a configure.ac or CMakeLists.txt file from scratch. They are copied and modified. Ineffective practices proliferate.

In these cases, and there are a few really common ones, I think Spack can help. It knows about compilers. It knows about MPIs. It knows about blas/lapacks. It knows about CMake, Autoconf. I don't see why it shouldn't also know about compiler's fortran name mangling.

markcmiller86 commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,142 @@
+       differential equations on block-structured adaptively refined
+       logically rectangular (i.e. Cartesian) grids."""
+
+    homepage = "https://anag-repo.lbl.gov/svn/Chombo/release"
```
I guess because my login there was cached, it worked for me. I've corrected to their main home pages instead of software download pages.

markcmiller86 commented:
@citibeth 

> That's the "easy" part. The "hard" part is using that information effectively in Chombo. I looked at Chombo's FortranNameMacro.H file, which is would would need to be patched/modified/updated. Here's the beginning...

In Make.defs, Chombo has a top-level make variable that controls this. We just need to know which of the available values (CH_FORT_UNDERSCORE, CH_FORT_UPPERCASE, CH_FORT_NOUNDERSCORE) to set it to. I think Chombo's Makefiles will handle the rest.

citibeth commented:
> But, they've got something thats been working for them that I am sure they
> would prefer not to have to go back and change now.
>
Maybe we can convince upstream developers these kinds of changes are worth
it, if it makes it easier for users to install their software.  I don't
think anyone *enjoys* manually installing a stack of 15 packages instead of
writing *spack install chombo*.

> In these cases, and there are a few really common ones, I think Spack can
> help. It knows about compilers. It knows about MPIs. It knows about
> blas/lapacks. It knows about CMake, Autoconf. I don't see why it shouldn't
> also know about compiler's fortran name mangling.
>
I tacitly agreed with you, I'm just disappointed we have to head in that
direction.  See #4136, maybe someone will add that feature.

citibeth commented:
Anyway... I think the right thing to do for now on the name mangling issue
for FASTmath is to punt.  Just leave things as they are, and add a comment
with URLs pointing to #4120 and #4136.

markcmiller86 commented:
I removed fastmath.tar.gz dummy package, derived from Package and switched to @BarrySmith method of handling install. I cannot test. spack install fastmath gives...

> ==> Error: There are no valid versions for spectrum-mpi that match ':'

adamjstewart commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -55,5 +55,11 @@ class Fastmath(MakefilePackage):
+    # Dummy install for now,  will be removed when metapackage is available
+    def install(self, spec, prefix):
+        # Prevent the error message
+        #      ==> Error: Install failed for xsdk.  Nothing was installed!
```
xsdk -> fastmath

Also, you'll need a URL now.

adamjstewart commented on var/spack/repos/builtin/packages/fastmath/package.py:
```diff
@@ -0,0 +1,64 @@
+    Solvers."""
+
+    homepage = "www.fastmath-scidac.org/"
+
```
Just need to add some random fake URL to this package now to get it working.

citibeth commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,140 @@
+
+    depends_on('blas')
+    depends_on('lapack')
+    depends_on('gmake', type='build')
```
@adamjstewart  Should we depend on `make` here?  (1) Wouldn't `MakefilePackage` do that anyway, (2) isn't it already assumed on the system?  Make is used to build a WHOLE LOT of packages, and I've never seen it as an explicit dependency.

Also, should `gmake` be upgraded to provide a virtual package called `make`?

adamjstewart replied:
MakefilePackage doesn't depend on make. It could, but there's no make package. Make is generally assumed to be on the system. However, if the package requires GNU Make specifically and won't work with other makes, then this dependency is correct. We can assume that some form of make is installed, but we can't assume that gmake is installed. At least, that's where I draw the line.

markcmiller86 replied:
Chombo requires GNU make

markcmiller86 replied:
On thing about @adamjstewart comments here..."make is generally assumed to be on the system...". That kinda sorta fits into this _deep dependency chains_ issue I've been muttering about in several posts/issues.

We need a way for each user of Spack to easily characterize what they think are things that are _generally assumed to be on the system_ such that Spack winds up dealing with dependency chains only as deeply as they expect.

I specifically set a gmake dependency in Chombo because I fear a system without gmake will have problems with it. Most makes are really gmake under the covers. But, this makes Chombo's gmake dependency explicit and I think its best

citibeth commented:
OK thanks.  So we just need an answer from Mark, and the make issue will be
resolved.  Once that's done, I'm ready to get this approved and merged.

On Tue, May 9, 2017 at 9:02 AM, Adam J. Stewart <notifications@github.com>
wrote:

> *@adamjstewart* commented on this pull request.
> ------------------------------
>
> In var/spack/repos/builtin/packages/chombo/package.py:
>
> > +
> +    variant('mpi', default=True, description='Enable MPI parallel support')
> +    variant('hdf5', default=True, description='Enable HDF5 support')
> +    variant('dims',
> +        default='3',
> +        values=('1', '2', '3', '4', '5', '6'),
> +        multi=False,
> +        description='Number of PDE dimensions [1-6]'
> +    )
> +
> +    patch('hdf5-16api.patch', when='@3.2', level=0)
> +    patch('Make.defs.local.template.patch', when='@3.2', level=0)
> +
> +    depends_on('blas')
> +    depends_on('lapack')
> +    depends_on('gmake', type='build')
>
> MakefilePackage doesn't depend on make. It could, but there's no make
> package. Make is generally assumed to be on the system. However, if the
> package requires GNU Make specifically and won't work with other makes,
> then this dependency is correct. We can assume that some form of make is
> installed, but we can't assume that gmake is installed. At least, that's
> where I draw the line.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/4120#discussion_r115482090>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd85RzBhS7nF4ZPer4k6V8Qc0lA2Uks5r4GPrgaJpZM4NQC1L>
> .
>

citibeth commented on var/spack/repos/builtin/packages/chombo/package.py:
```diff
@@ -0,0 +1,140 @@
+
+    def build(self, spec, prefix):
+        with working_dir('lib'):
+            make('all')
```
We have determiend that Chombo requires GNU make.  But won't this `make('all')` line use whatever system make is lying around?  @adamjstewart ???

adamjstewart replied:
Yes, good catch! You'll have to use `gmake()` instead of `make()`. These are defined [here](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/build_environment.py#L394).

markcmiller86 replied:
Hmm. Good question. Does that need to be something like `gmake('all')` or maybe `gm = which(gmake) gm('all')`

adamjstewart replied:
`gmake('all')` should work fine.

citibeth commented:
> I specifically set a gmake dependency in Chombo because I fear a system
> without gmake will have problems with it. Most makes are really gmake under
> the covers. But, this makes Chombo's gmake dependency explicit and I think
> its best
>
I'm not doubting your choice here.  But the command `make('install')` will
unfortunately not run the GMake on which you declared a dependency, it will
run the system make.  Adam will have a good solution to that issue.

citibeth commented:
Will that use the GMake this package depends on, or just one Spack found
lying around the system?

On Tue, May 9, 2017 at 1:50 PM, Adam J. Stewart <notifications@github.com>
wrote:

> *@adamjstewart* commented on this pull request.
> ------------------------------
>
> In var/spack/repos/builtin/packages/chombo/package.py:
>
> > +                '^\s*#\s*HDFINCFLAGS\s*=.*',
> +                'HDFINCFLAGS = -I%s' % spec['hdf5'].prefix.include)
> +            defs_file.filter(
> +                '^\s*#\s*HDFLIBFLAGS\s*=.*',
> +                'HDFLIBFLAGS = %s' % spec['hdf5'].libs.ld_flags)
> +            if '+mpi' in spec:
> +                defs_file.filter(
> +                    '^\s*#\s*HDFMPIINCFLAGS\s*=.*',
> +                    'HDFMPIINCFLAGS = -I%s' % spec['hdf5'].prefix.include)
> +                defs_file.filter(
> +                    '^\s*#\s*HDFMPILIBFLAGS\s*=.*',
> +                    'HDFMPILIBFLAGS = %s' % spec['hdf5'].libs.ld_flags)
> +
> +    def build(self, spec, prefix):
> +        with working_dir('lib'):
> +            make('all')
>
> Yes, good catch! You'll have to use gmake() instead of make(). These are
> defined here
> <https://github.com/LLNL/spack/blob/develop/lib/spack/spack/build_environment.py#L394>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/4120#discussion_r115555796>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd6k1HFPSsCFJS_kzTMM1ksgaPBWAks5r4KdUgaJpZM4NQC1L>
> .
>

adamjstewart commented:
It will use the GMake built by Spack, as Spack adds all build dependencies to the PATH. Although I'm not quite sure _when_ it does this vs. when it loads `build_environment`. If you print `gmake.path` it will show which one it's using.

markcmiller86 commented:
So, just to confirm, I am thinking these CI failures have something to do with something that doesn't involve me or anything in this PR. Is that right? Or, am I missing something? I don't see how any of the recent changes I've made could have caused these changes in CI behavior.

adamjstewart commented:
@markcmiller86 See #4180. I restarted Travis so it should be good now.
